Core enabled hw iso onwards

deepakala-k opened the pull request:
The following commits are cherry picked
[redfish-core: Core: Enabled the isolation (aka guard) feature](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/1780452abef6b9e167ffac8c27d6cbdd0be7934f)

[redfish-core: Memory: Enabled the isolation (aka guard) feature](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/70c213769052d24244ce58e9836f6f45cedf474c)

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+        // De-isolate the given resource
+        crow::connections::systemBus->async_method_call(
+            [aResp,
+             resourceIsolatedHwEntry](const boost::system::error_code ec1,
```
```
const boost::system::error_code& ec1
```

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+                             dbusError->message);
+
+            if (strcmp(dbusError->name, "xyz.openbmc_project.Common.Error."
+                                        "NotAllowed") == 0)
```
Keep a single string and use std::string_view comparison
```
            if (std::string_view("xyz.openbmc_project.Common.Error.NotAllowed") == dbusError->name)
```

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+                messages::propertyNotWritable(aResp->res, "Entry");
+            }
+            else if (strcmp(dbusError->name, "xyz.openbmc_project.Common.Error."
+                                             "Unavailable") == 0)
```
Keep a single string and use std::string_view comparison
```
            else if (std::string_view("xyz.openbmc_project.Common.Error.Unavailable) == dbusError->name)
```

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+ *         given "Enabled" property value.
+ */
+inline void
+    processHardwareIsolationReq(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
```
`aResp` -> `asyncResp`

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+    },
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetSubTreePaths",
```
```
dbus::utility::getSubTreePaths()
```

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+    crow::connections::systemBus->async_method_call(
+        [aResp, resourceName, resourceId,
+         enabled](boost::system::error_code ec,
+                  const std::vector<std::string>& objects) {
```
```
boost::system::error_code& ec,
const dbus::utility::MapperGetSubTreePathsResponse& objects)
```

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+        // Get the HardwareIsolation DBus name
+        crow::connections::systemBus->async_method_call(
+            [aResp, resourceObjPath, enabled, resourceName,
+             resourceId](const boost::system::error_code ec1,
```
```
const boost::system::error_code& ec1
```

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+        },
+            "xyz.openbmc_project.ObjectMapper",
+            "/xyz/openbmc_project/object_mapper",
+            "xyz.openbmc_project.ObjectMapper", "GetObject",
```
```
dbus::utility::getDbusObject(
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -892,6 +897,13 @@ inline void
+        "/redfish/v1/Systems/system/Processors/" + processorId;
+    asyncResp->res.jsonValue["SubProcessors"] = {
+        {"@odata.id", "/redfish/v1/Systems/system/Processors/" + processorId +
+                          "/SubProcessors"}};
```
```
    asyncResp->res.jsonValue["SubProcessors"]["@odata.id"] =
         boost::urls::format("/redfish/v1/Systems/system/Processors/{}/SubProcessors", processorId);
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -892,6 +897,13 @@ inline void
 {
+    asyncResp->res.jsonValue["@odata.type"] = "#Processor.v1_12_0.Processor";
+    asyncResp->res.jsonValue["@odata.id"] =
+        "/redfish/v1/Systems/system/Processors/" + processorId;
```
```
boost::urls::format("/redfish/v1/Systems/system/Processors/{}",  processorId);
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+{
+    crow::connections::systemBus->async_method_call(
+        [processorId, aResp, handler{std::forward<Handler>(handler)}](
+            const boost::system::error_code ec,
```
```
const boost::system::error_code& ec
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+    crow::connections::systemBus->async_method_call(
+        [processorId, aResp, handler{std::forward<Handler>(handler)}](
+            const boost::system::error_code ec,
+            const std::vector<std::string>& subTreePaths) {
```
```
const dbus::utility::MapperGetSubTreePathsResponse& subTreePaths) {
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+    },
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetSubTreePaths",
```
```
dbus::utility::getSubTreePaths()
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+    aResp->res.jsonValue["Status"]["Health"] = "OK";
+
+    crow::connections::systemBus->async_method_call(
+        [objPath, aResp](const boost::system::error_code ec,
```
```
const boost::system::error_code& ec
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+            for (const auto& [interface, properties] : interfaces)
+            {
+                if (interface == "xyz.openbmc_project.State."
+                                 "Decorator.OperationalStatus")
```
Keep a single string
```
                if (interface == "xyz.openbmc_project.State.Decorator.OperationalStatus")
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+}
+
+inline void
+    getCpuCoreDataByService(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
```
`aResp` -> `asyncResp`

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+        }
+    },
+        service, "/xyz/openbmc_project/inventory",
+        "org.freedesktop.DBus.ObjectManager", "GetManagedObjects");
```
```
dbus::utility::getManagedObjects()
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+        [aResp, processorId, coreId](const std::string& cpuPath) {
+        crow::connections::systemBus->async_method_call(
+            [aResp, processorId, coreId](
+                const boost::system::error_code ec,
```
```
const boost::system::error_code& ec
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+                const boost::container::flat_map<
+                    std::string, boost::container::flat_map<
+                                     std::string, std::vector<std::string>>>&
+                    subtree) {
```
```
const dbus::utility::MapperGetSubTreeResponse& subtree) {
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+        },
+            "xyz.openbmc_project.ObjectMapper",
+            "/xyz/openbmc_project/object_mapper",
+            "xyz.openbmc_project.ObjectMapper", "GetSubTree", cpuPath, 0,
```
```
dbus::utility::getSubTree()
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+                    std::string("/redfish/v1/Systems/system/Processors/")
+                        .append(processorId)
+                        .append("/SubProcessors/")
+                        .append(coreId);
```
This needs to be clearer 
```
                asyncResp->res.jsonValue["@odata.id"] = boost::urls::format(
                            "/redfish/v1/Systems/system/Processors/{}/SubProcessors/{}", 
                            processorId, coreId);
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+{
+    auto callback = [aResp, processorId](const std::string& cpuPath) {
+        crow::connections::systemBus->async_method_call(
+            [processorId, aResp](const boost::system::error_code ec,
```
```
const boost::system::error_code& ec
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+    auto callback = [aResp, processorId](const std::string& cpuPath) {
+        crow::connections::systemBus->async_method_call(
+            [processorId, aResp](const boost::system::error_code ec,
+                                 const std::vector<std::string>& subTreePaths) {
```
```
const dbus::utility::MapperGetSubTreePathsResponse& subTreePaths) {
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+                "#ProcessorCollection.ProcessorCollection";
+            aResp->res.jsonValue["@odata.id"] =
+                "/redfish/v1/Systems/system/Processors/" + processorId +
+                "/SubProcessors";
```
```
            aResp->res.jsonValue["@odata.id"] =
                boost::urls::format("/redfish/v1/Systems/system/Processors/{}/SubProcessors", processorId)
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+            members = nlohmann::json::array();
+            std::string subProcessorsPath =
+                "/redfish/v1/Systems/system/Processors/" + processorId +
+                "/SubProcessors/";
```
```
std::string subProcessorsPath =
         boost::urls::format("/redfish/v1/Systems/system/Processors/{}/SubProcessors", processorId);
```
(Note the tailend "/" is handled when it is added)

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+                members.push_back(
+                    {{"@odata.id", subProcessorsPath +
+                                       sdbusplus::message::object_path(corePath)
+                                           .filename()}});
```
```
            nlohmann::json::object_t item;
            item["@odata.id"] = boost::urls::format("{}/{}", subProcessorsPath,
                                       sdbusplus::message::object_path(corePath).filename());
               members.emplace_back( std::move(item) );
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,276 @@ inline void
+        },
+            "xyz.openbmc_project.ObjectMapper",
+            "/xyz/openbmc_project/object_mapper",
+            "xyz.openbmc_project.ObjectMapper", "GetSubTreePaths", cpuPath, 0,
```
```
dbus::utility::getSubTreePaths()
```

baemyung commented on redfish-core/lib/processor.hpp:
```diff
@@ -49,6 +50,10 @@ constexpr std::array<std::string_view, 2> processorInterfaces = {
 
+// Interfaces which imply a D-Bus object represents a Processor Core
+constexpr std::array<const char*, 1> procCoreInterfaces = {
+    "xyz.openbmc_project.Inventory.Item.CpuCore"};
```
Can we use `std::string_view` here instead of `const char*`?

baemyung commented on redfish-core/lib/memory.hpp:
```diff
@@ -43,6 +44,10 @@
 
+// Interfaces which imply a D-Bus object represents a Memory
+constexpr std::array<const char*, 1> dimmInterfaces = {
+    "xyz.openbmc_project.Inventory.Item.Dimm"};
```
It would be better to use `std::string_view` instead of `const char*` for the consistency throughout bmcweb code base

deepakala-k replied:
done

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+                         const dbus::utility::ManagedObjectType& dbusData) {
+        if (ec)
+        {
+            BMCWEB_LOG_DEBUG("DBUS response error, ec: {}", ec.value());
```
Should this be BMCWEB_LOG_ERROR()?

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+                if (interface == "xyz.openbmc_project.State."
+                                 "Decorator.OperationalStatus")
+                {
+                    for (const auto& [proName, proValue] : properties)
```
nit: I think propName and propValue would be better names for these variables.

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+                            const bool* value = std::get_if<bool>(&proValue);
+                            if (value == nullptr)
+                            {
+                                messages::internalError(aResp->res);
```
Should there be a BMCWEB_LOG_ERROR() here to help with debug?

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+                            const bool* value = std::get_if<bool>(&proValue);
+                            if (value == nullptr)
+                            {
+                                messages::internalError(aResp->res);
```
Should there be a BMCWEB_LOG_ERROR() here to help with debug?

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+                                std::get_if<std::string>(&proValue);
+                            if (prettyName == nullptr)
+                            {
+                                messages::internalError(aResp->res);
```
Should there be a BMCWEB_LOG_ERROR() here to help with debug?

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+{
+    BMCWEB_LOG_DEBUG("Get available system sub processor resources.");
+
+    auto callback = [aResp, processorId, coreId](const std::string& cpuPath) {
```
Long lambda should be factored out into named function per upstream coding style. (L1137-L1200)

deepakala-k replied:
Will address it as part of the new JIRA ticket

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+                const boost::container::flat_map<
+                    std::string, boost::container::flat_map<
+                                     std::string, std::vector<std::string>>>&
+                    subtree) {
```
Long lambda should be factored out into named function per upstream coding style. (L1144-L1194)

deepakala-k replied:
Will address it as part of the new JIRA ticket

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+                    subtree) {
+            if (ec)
+            {
+                BMCWEB_LOG_DEBUG("DBUS response error, ec: {}", ec.value());
```
Should this be BMCWEB_LOG_ERROR() instead?

jeaaustx commented on redfish-core/lib/processor.hpp:
```diff
@@ -975,6 +987,275 @@ inline void
+    auto callback = [aResp, processorId](const std::string& cpuPath) {
+        crow::connections::systemBus->async_method_call(
+            [processorId, aResp](const boost::system::error_code ec,
+                                 const std::vector<std::string>& subTreePaths) {
```
Long lambda, move into named function.

jeaaustx reviewed (changes requested):
The Implement SubProcessor commit rebase for 1110 adds some of these functions and has been updated with the upstream code style. This commit might be better to go on top of that one.

https://github.com/ibm-openbmc/bmcweb/pull/988

deepakala-k commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+                                            resourceName, resourceId);
+        }
+        else if (strcmp(dbusError->name, "xyz.openbmc_project.Common.Error."
+                                         "TooManyResources") == 0)
```
done

deepakala-k commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+#pragma once
+
```
done

deepakala-k commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+        crow::connections::systemBus->async_method_call(
+            [aResp, resourceObjPath, enabled, resourceName,
+             resourceId](const boost::system::error_code ec1,
+                         const dbus::utility::MapperGetObject& objType) {
```
Will address as part of new JIRA

deepakala-k commented on redfish-core/lib/processor.hpp:
```diff
@@ -892,6 +897,13 @@ inline void
                      const std::string& objectPath,
                      const dbus::utility::MapperServiceMap& serviceMap)
 {
+    asyncResp->res.jsonValue["@odata.type"] = "#Processor.v1_12_0.Processor";
```
I will follow up and address this as part of new JIRA

deepakala-k commented on redfish-core/lib/memory.hpp:
```diff
@@ -24,6 +24,7 @@
 #include "utils/collection.hpp"
 #include "utils/dbus_utils.hpp"
 #include "utils/hex_utils.hpp"
+#include "utils/hw_isolation.hpp"
```
Some of the required changes are missing due to rebasing. I will address it. Thank you

deepakala-k commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,327 @@
+        crow::connections::systemBus->async_method_call(
+            [aResp,
+             resourceIsolatedHwEntry](const boost::system::error_code ec1,
+                                      const sdbusplus::message::message& msg) {
```
Will address it later as part of the new JIRA

deepakala-k commented:
> The Implement SubProcessor commit rebase for 1110 adds some of these functions and has been updated with the upstream code style. This commit might be better to go on top of that one.
> 
> #988

Yeah.. I have rebased my commit on top of it. Thanks

deepakala-k commented:
I have addressed the comments that are small. Will do the other changes shortly as part of other JIRA ticket as it is difficult to resolve conflicts on the other commits that I have in pipeline

baemyung commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,337 @@
+            }
+            else if (strcmp(dbusError->name,
+                            "xyz.openbmc_project.Common.Error.Unavailable") ==
+                     0)
```
The use of `strcmp()`  can also be replaced using `std::string_view()` comparison like L61 -- probably in the next Jira item

```
            if (std::string_view("xyz.openbmc_project.Common.Error.NotAllowed") == dbusError->name)
```

baemyung commented on redfish-core/lib/memory.hpp:
```diff
@@ -809,6 +871,33 @@ inline void getDimmData(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+ *         "xyz.openbmc_project.Object.Enable::Enabled" dbus property.
+ */
+
+inline void patchMemberEnabled(const std::shared_ptr<bmcweb::AsyncResp>& resp,
```
Please use `asynResp` for all occurrences

deepakala-k replied:
done

baemyung reviewed (approved):
I think this seems good to me functionally.

A few more comments are added.
As many of the comments would be considered as a next Jira item,  they can also be done at the time.

deepakala-k commented on redfish-core/lib/memory.hpp:
```diff
@@ -713,6 +718,53 @@ inline void getDimmPartitionData(std::shared_ptr<bmcweb::AsyncResp> asyncResp,
+{
+    crow::connections::systemBus->async_method_call(
+        [asyncResp{std::move(asyncResp)}](
+            const boost::system::error_code ec,
```
done
